### ISSUE TITLE
Binding Datagrid's SelectedItem

### ISSUE THREAD

**Issue by felipeleon73** [elmish/Elmish.WPF#77]
How to bind two way on SelectedItem?
I have the ItemsSource bound with a Binding.subBindingSeq.
Now SelectedItem from Datagrid return a ViewModel<obj,obj>.
I can read the value of the item model with CurrentModel, but how set the relative ViewModel in the setter?

**Comment by 2sComplement** [comment]
Is this related to #63?

**Comment by felipeleon73** [comment]
I don't think...
My issue is that i can't set the "ViewModel selected item" by the "model slected item".
In #63 hare used SelectedValue+SelectedValuePath, but i use Syncfusion SfDatagrid that seems to haven't this properties

**Comment by cmeeren** [comment]
@felipeleon73 If you can provide a minimal repro solution, I'm happy to take a look at it! :)

**Comment by felipeleon73** [comment]
This is my [repo](https://github.com/felipeleon73/Elmish.WPF.Collection.Test).
The problem is in Program.fs in
```
let bindings model dispatch =
  [
    "Count" |> Binding.oneWay (fun m -> m.Count)
    "Selected" |>  Binding.twoWay 
                    (fun m -> unbox m.Selected) //<-----------here is the issue! m.Selected is a Model, expect a ViewModel
                    (fun (v:ViewModel<_,_>) m -> Update v.CurrentModel)
    "Items" |> Binding.subBindingSeq
        id
        (fun m -> m.Items)
        (fun c -> c.Id)
        itemBindings
  ]
```

**Comment by JDiLenarda** [comment]
Not sure if subBindingSeq is involved in your problem. You’re not supposed to work with `ViewModel<_,_>` as such. Try this (untested) : 

```F#
"Selected" |> Bindings.twoWay
          (fun m -> Option.toObj m.Selected)
          (fun v _ -> Update (Option.ofObj v))
```

There, `v` should be a Nullable of Item.

**Comment by cmeeren** [comment]
@JDiLenarda is partly right. AFAIK it should look like this:

```f#
"Selected" |>  Binding.twoWay 
  (fun m -> m.Selected |> Option.map box |> Option.toObj)
  (fun v m -> v |> Option.ofObj |> Option.map unbox |> Select )
```

However, this throws on the last line when selecting an item, because `v` is of type `ViewModel`. I'm not immediately sure why. I'll need to investigate. Busy the next few days if anyone else wants to have a look.

**Comment by cmeeren** [comment]
Nevermind, makes sense now. The items are indeed `ViewModel<_,_>` instances because of the other binding, `"Items" |> Binding.subBindingSeq ...`. However, that means that you can't bind directly to a `SelectedItem` property like this. But that's poor practice anyway, because you get duplicated state in your model (if you update the item that is selected, you'll have to update it both in `Selected` and in `Items`). The solution is to use `SelectedIndex`, which the control has.

**Comment by felipeleon73** [comment]
Thanx. 
The Selectedindex trick works perfectly in most cases...  unless the DataGrid is filtered or sorted.
You believe it is possible to implement a mechanism to set and retrive del selected item using the unique key shared by the model and the view model?

**Comment by cmeeren** [comment]
If you sort or filter the DataGrid, I would guess you need to do it in the Elmish model.

Other than that, I don't have any good ideas at the moment if the control does not support `SelectedValuePath` or another means of specifying the model member you wish to use when selecting.

**Comment by cmeeren** [comment]
Hmm... There might be a way to use the unique key as you suggest. I need to think a bit. Busy these next days though, so can't promise when I'll get to it.

**Comment by cmeeren** [comment]
@felipeleon73, I think I found a way to achieve what you want. Please check out the `subModelSelectedItem` branch and use the new `Binding.subModelSelectedItem`. It takes care of converting from/to the model IDs, so the Elmish usage is very simple.

I have also added a dedicated sample for this binding.

Note that (as the new binding function documentation describes) it's not as type-safe as a normal `twoWay` binding with `SelectedValue` or `SelectedIndex`, so those are still preferred if possible.

**Comment by felipeleon73** [comment]
Perfect!
Thank you very much!

**Comment by felipeleon73** [comment]
Really really great solution!
Works like a charm!

**Comment by cmeeren** [comment]
Will be released as 2.0.0-beta-11, currently in the pipeline.
